kliff/dataset/weight.py

```diff
@@ -0,0 +1,167 @@
+
+    This class mimics the behavior  provided by the ``residual_data`` argument from the
+    previous version.
+    Subclass can implement
+        - compute_weight
+
+    Args:
+        config_weight: the weight for the entire configuration
+        energy_weight: the weight for the energy
+        forces_weight: the weight for the forces
+        stress_weight: the weight for the stress
+    """
+
+    def __init__(
+        self,
+        config_weight: float = 1.0,
+        energy_weight: float = 1.0,
+        forces_weight: float = 1.0,
+        stress_weight: float = 1.0,
+    ):
+        self._config_weight = config_weight  # weight for the entire configuration
+        self._energy_weight = energy_weight
+        self._forces_weight = forces_weight
+        self._stress_weight = stress_weight
+
+    @staticmethod
+    def compute_weight(config):
+        pass
+
+    @property
+    def config_weight(self):
+        return self._config_weight
+
+    @property
+    def energy_weight(self):
+        return self._energy_weight
+
+    @property
+    def forces_weight(self):
+        return self._forces_weight
+
+    @property
+    def stress_weight(self):
+        return self._stress_weight
+
+
+default_weight_params = {
+    "energy_weight_params": [1.0, 0.0],
+    "forces_weight_params": [1.0, 0.0],
+    "stress_weight_params": [1.0, 0.0],
+}
+
+
+class NonuniformWeight(Weight):
+    r"""
+    Non-uniform weight that is computed from the data. The calculation follows Lenosky et
+    al. (1997), with some modification in notation,
+
+    ..math:
+       \frac{1}{w_m^2} = c_1^2 + c_2^2 \Vert f_m \Vert ^2.
+
+    Args:
+        config_weight: the weight for the entire configuration
+        weight_params: a dictionary containing parameters c1 and c2 for calculating the
+            weight of each property. The supported key value pairs are:
+            - energy_weight_params: float or array-like (default: [1.0, 0.0])
+            - forces_weight_params: float or array-like (default: [1.0, 0.0])
+            - stress_weight_params: float or array-like (default: [1.0, 0.0])
+            If a float is given, this number will be used to set c1, while c2 is set to
+            zero. If an array-like with 2 elements is given, it should contain c1 as the
+            first element and c2 as the second element.
+    """
+
+    def __init__(
+        self, config_weight: float = 1.0, weight_params: Optional[Dict[str, Any]] = None
+    ):
+        self._config_weight = config_weight
+        # Initiate the weight values. They will be changed latter.
+        self._energy_weight = 0.0
+        self._forces_weight = 0.0
+        self._stress_weight = 0.0
+
+        self._weight_params = self._check_weight_params(
+            weight_params, default_weight_params
+        )
+
+    def compute_weight(self, config):
+        """
+        Compute the weights of the energy, forces, and stress data.
+        """
+        energy = config._energy
```

Comment:
energy = config.energy should be good. Same for forces and stress.

Reply:
I did this because if I call `config.stress` when the dataset doesn't contain stress, it raises an error. I can use try catch to overcome this.

Reply:
Ah, right. Then let's keep the underscore version for now. Will probably update the Configuration class later.